controller/service.go

```diff
@@ -266,3 +277,17 @@ func isEqualIPs(ipsA, ipsB []net.IP) bool {
 	})
 	return reflect.DeepEqual(ipsA, ipsB)
 }
+
+// getIPHostnameRelation creates a map with the ips that have a hostname set externally.
+func getIPHostnameRelation(lbHostnames []string, lbIPs []net.IP) map[string]string {
+	ipHostname := make(map[string]string)
```

Comment:
We can fill the map while we iterate the ingress structure (either in the loop above, or by re iterating it as it's gonna be a few elements anyway). By doing that we have more clear the relationship between ip and the name, while here we need to reason about the fact that we must process in the same order or if there's an ip in the list that does not have an hostname associated.

Reply:
Yes, the idea was to iterate the ingress structure again, but everything is deleted in line 75 when a service has no lb ips previously assigned (first time allocations). Therefore, the hostnames are lost.

Which loop are you referring to exactly?

Reply:
What I mean is, you can build the map of the desired association where you collect `lbHostnames`, and use it later.

Reply:
how can I build the map there if there are no ips (except when there is a reassignment)? The key of the map is the IP and the value the hostname. Sorry, I don't understand.

Reply:
Sorry, let's stop for a sec. I might have misunderstood the whole thing.

My thought was that in lb services, any external controller would associate an hostname to an ip that is being set by metallb. So the flow would've been: metallb sets the ip, the controller associates the hostname to the ip.

Looking at the api docs they seem mutually exclusives:
```
// LoadBalancerIngress represents the status of a load-balancer ingress point:
// traffic intended for the service should be sent to an ingress point.
type LoadBalancerIngress struct {
	// IP is set for load-balancer ingress points that are IP based
	// (typically GCE or OpenStack load-balancers)
	// +optional
	IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`

	// Hostname is set for load-balancer ingress points that are DNS based
	// (typically AWS load-balancers)
	// +optional
	Hostname string `json:"hostname,omitempty" protobuf:"bytes,2,opt,name=hostname"`

	// Ports is a list of records of service ports
	// If used, every port defined in the service should have an entry in it
	// +listType=atomic
	// +optional
	Ports []PortStatus `json:"ports,omitempty" protobuf:"bytes,4,rep,name=ports"`
}
```

So, if that is the case, the logic is even simpler. Sorry again, I probably misunderstood the whole thing.